src/tree/mod.rs

```diff
@@ -700,14 +700,14 @@ impl<T> Tree<T> {
     {
         let lower_upper_test = self.find_subtree_root_between_ids(first_id, second_id)
             .map(|_| (first_id, second_id))
-            .or(self.find_subtree_root_between_ids(second_id, first_id)
+            .or_else(|| self.find_subtree_root_between_ids(second_id, first_id)
```

Comment:
That looks totally fine and that evaluation thing makes sense. I would go for it.